### PR TITLE
GTEST: Mock protocol selection tests

### PR DESCRIPTION
**Pull request by iyastreb**
## What
Mock infrastructure and reference test for protocol selection

## Why ?
To enable writing tests for protocol selection, that are executed on different HW

### CODE REVIEW

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+            ucp_proto_query_attr_t proto_attr;
+            int valid = ucp_proto_select_elem_query(worker, &select_elem,
+                                                    range_start, &proto_attr);
+            range_end = proto_attr.max_msg_length;
+            if (valid) {
+            char buf[1024];
+                sprintf(buf, "  %zu-%zu desc: %s, config: %s", range_start,
+                range_end, proto_attr.desc, proto_attr.config);
+                UCS_TEST_MESSAGE << buf;
+            }
```
Comment (ivankochin) [inline test/gtest/ucp/test_ucp_proto_select.cc:363]:
```suggestion
            if (ucp_proto_select_elem_query(worker, &select_elem,
                                            range_start, &proto_attr) {
                UCS_TEST_MESSAGE << range_start <<  "-" << proto_attr.max_msg_length << " dest: " << proto_attr.desc << ", config: " << proto_attr.config;
            }
            range_end = proto_attr.max_msg_length;
```

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    static void
+    check_proto_select_elem(ucp_worker_h worker,
+                            const ucp_proto_select_elem_t &select_elem,
+                            const proto_select_data_vec_t &data)
+    {
+        dump_select_elem(worker, select_elem);
+        for (auto &it : data) {
+            size_t start = string_to_memunits(it.range_start);
+            EXPECT_TRUE(select_elem_match(worker, select_elem, it, start));
+            /* As we cannot get range_start directly, we assert that protocol
```
Comment (ivankochin) [inline test/gtest/ucp/test_ucp_proto_select.cc:376]: Why we cannot get range_start directly?

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+    static ucs_status_t iface_open_mock(uct_md_h md, uct_worker_h worker,
+                                        const uct_iface_params_t *params,
+                                        const uct_iface_config_t *config,
+                                        uct_iface_h *iface_p)
+    {
+        mock_component *self = find(md->component);
+        uct_tl_t *tl         = self->find_tl(params);
+        ucs_status_t status  = mock_t::actual_call(&tl->iface_open, md, worker,
+                                                   params, config, iface_p);
+        if (status == UCS_OK) {
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_proto_select.cc:191]: Minor: invert branches in all these places. Better to have:
```
if (status != UCS_OK) {
    return status;
}

...
...
return UCS_OK;
```

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+#include <ucp/proto/proto_select.inl>
+#include <ucs/sys/topo/base/topo.h>
+}
+
+struct mock_t {
+public:
+    typedef void (*func_t)(void);
+    typedef std::map<func_t *, std::unique_ptr<mock_t>> mock_map_t;
+
+    ~mock_t()
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_proto_select.cc:23]: Can move to `private`, under the ctor?

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+    static void
+    check_ep_config(entity &e, const proto_select_data_vec_t &data,
+                    ucp_proto_select_key_t key)
+    {
+        ucp_ep_config_t *config = ucp_worker_ep_config(e.worker(), 0);
+        check_proto_select(e.worker(), config->proto_select, data, key);
+    }
+
+    static void
+    check_rkey_config(entity &e, const proto_select_data_vec_t &data,
```
Comment (ivankochin) [inline test/gtest/ucp/test_ucp_proto_select.cc:292]: Never called?

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+        }
+
+        return status;
+    }
+
+    uct_component_h                           m_component;
+    uint64_t                                  m_flags_orig;
+    std::map<uct_base_iface_t *, std::string> m_iface_names;
+    std::map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
```
Comment (ivankochin) [inline test/gtest/ucp/test_ucp_proto_select.cc:243]: Is there a necessity to be able to set different values for different transport devices? WDYTA removing all this maps and replacing them by one instance of `iface_attr_func_t` and `perf_attr_func_t` which would be used to mock all the corresponding calls?

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+extern "C" {
+#include <ucp/core/ucp_ep.inl>
+#include <uct/base/uct_component.h>
+#include <uct/base/uct_iface.h>
+#include <uct/base/uct_md.h>
+#include <ucp/proto/proto_select.inl>
+#include <ucs/sys/topo/base/topo.h>
+}
+
+struct mock_t {
```
Comment (ivankochin) [inline test/gtest/ucp/test_ucp_proto_select.cc:18]: If I understand correctly both `mock_t` and `mock_component` entities are used for two purposes where first is storage of mock and origin pointer or component and operations under them and another are their static interfaces which are supposed to store some global containers with `mock_t` or `mock_component` and making some operations with them.

So I want to propose to split these to API to two different entities: make `mock_t` and `mock_component` APIs as minimalistic as possible and move all the containers and operations with outside of the classes (maybe create separate workspace for them). To incapsulate map containers we can use unnamed namespaces.

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+        g_mock_map.clear();
+    }
+
+    template <typename Fn>
+    static mock_t *setup(Fn *orig_ptr, Fn mock)
+    {
+        func_t *ptr = reinterpret_cast<func_t *>(orig_ptr);
+        auto it = g_mock_map.find(ptr);
+        if (it == g_mock_map.end()) {
+            mock_t * new_mock = new mock_t(ptr, reinterpret_cast<func_t>(mock));
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_proto_select.cc:39]: There's `make_unique` since c++14. Currently we use c++11.
Let's see if [this one](https://github.com/openucx/ucx/pull/10018) gets merged soon. If yes - can consider using `make_unique` here then.

Comment (gleon99) [reply test/gtest/ucp/test_ucp_proto_select.cc:39]: Ignore.

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 5 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:292]: Yes, it's never called so far.
To me the goal of this commit is to provide meaningful infra for writing mock tests for protocol selection. Currently I have just one test that is just an example. I think API for checking rkey config would be useful in the future

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+        g_component_map.clear();
+        mock_t::cleanup();
+    }
+
+    uct_component_t * operator->() const
+    {
+        return m_component;
+    }
+
+    void set_mock_perf_attr(const std::string &iface_name, perf_attr_func_t cb)
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:116]: Yes, so far it's never used.
I thought of this function as an API for extended protocol selection tests. But if we stick to mocking `iface_attr_t`, then probably this one we can remove.

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+    {
+        func_t *ptr = reinterpret_cast<func_t *>(orig_ptr);
+        auto it = g_mock_map.find(ptr);
+        if (it == g_mock_map.end()) {
+            mock_t * new_mock = new mock_t(ptr, reinterpret_cast<func_t>(mock));
+            g_mock_map[ptr] = std::unique_ptr<mock_t>(new_mock);
+            return new_mock;
+        }
+
+        return it->second.get();
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:44]: Ok, there are 2 points here:

1. Why do we mock the same functions multiple times?
This happens unintentionally, just because UCT creates multiple interfaces of the same transport type. Since we mock iface open with `iface_open_mock`, we need to replace iface_query with mock. But if we create more than one iface of the same type - this situation happens. It is not an error in this sense.

2. Why don't we overwrite?
Yes, we can overwrite, it looks logical

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+
+    template <typename Fn, typename... Args>
+    static ucs_status_t actual_call(Fn *orig_ptr, Args&&... args)
+    {
+        mock_t *mock = find(reinterpret_cast<func_t *>(orig_ptr)); 
+        Fn orig_func = mock ? reinterpret_cast<Fn>(mock->m_orig) : *orig_ptr;
+        return orig_func(std::forward<Args>(args)...);
+    }
+
+    static mock_t *find(func_t *orig_ptr)
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:55]: ok, why not

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 3 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:191]: ok

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+        return status;
+    }
+
+    uct_component_h                           m_component;
+    uint64_t                                  m_flags_orig;
+    std::map<uct_base_iface_t *, std::string> m_iface_names;
+    std::map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
+
+    typedef std::map<uct_component_h,
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:245]: done

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 2 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:376]: Because this field does not exist neither in `ucp_proto_query_attr_t` nor in `ucp_proto_threshold_elem_t`. We kind of keep only the upper limit, and lower limit we deduce from the array: e.g. check `ucp_proto_select_elem_info`, `ucp_proto_select_write_info`

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 1 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:355]: ok, applied with some minor changes

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+            }
+        } while (range_end != SIZE_MAX);
+    }
+
+    static void
+    check_proto_select_elem(ucp_worker_h worker,
+                            const ucp_proto_select_elem_t &select_elem,
+                            const proto_select_data_vec_t &data)
+    {
+        dump_select_elem(worker, select_elem);
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:364]: Yes, we need something like that, will be fixed
This dumping is very useful when writing new tests. Maybe dumping on error is sufficient for this purpose too

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 6 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc]: It's hard to say whether it's really needed, it depends what kind of tests we wish to create.
This gives us possibility to mock multiple transport/ifaces within the same test, which I think might be useful to:
1. Test proto selection between multiple ifaces (e.g. when we switch from one iface to another)
2. Test lane selection
I thought it's important to have this possibility (especially if it does not come with huge complexity). Only if we are 100% sure that those tests will never be needed - then we can simplify and stick to a single instance.

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+    ~mock_component()
+    {
+        m_component->flags = m_flags_orig;
+    }
+
+    static mock_component &setup(const std::string &name)
+    {
+        uct_component_h component = find_uct_component(name);
+        if (nullptr == component) {
+            UCS_TEST_SKIP_R("Component " + name + " not found");
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:87]: Sure, if we consider simplest use case with single transport - this skip makes no sense.
I was thinking of cases with multiple transports specified, like implicit "all" or specific "self,shmem,tcp,ib".
In this case - is it guaranteed that ALL those transports exist?
I was not sure, therefore this relaxed skip.

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 7 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc]: I refactored the code taking into account your feedback (also from our discussion). Key changes:
1. Split mock_t into outer mock { and inner struct mock_func }
2. Removed static global map - in fact we can instantiate mock per test, no need in global instance
3. Moved mock into test_helpers
4. Split mock_component into 2 entities:
- mock_component_test - This is the base class for component mocking, containing mock instance + implements global singleton map(), and setup_mock functions
- mock_component - this is API and infra for individual component mocking

Hope that helps

File: test/gtest/ucp/test_ucp_proto_select.cc (hunk 2 above)
Comment (ivankochin) [reply test/gtest/ucp/test_ucp_proto_select.cc:308]: As I can see, `select_elem_match` for `start - 1` would always return `false` since we compare `data.range_end` for current proto with `proto_attr.max_msg_length` which obviously would be different at `start - 1` point. So even if there are the same `desc` and `config` it would return false. AFAIU it would never happen now since `select_elem::thresholds` elements are merged together if same proto is selected, but I am not sure whether it is a good idea to rely on that.

File: test/gtest/ucp/test_ucp_proto_select.cc [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+mock_component &mock_component_test::setup_mock(const std::string &name)
+{
+    uct_component_h component = find_uct_component(name);
+    if (nullptr == component) {
+        UCS_TEST_SKIP_R("Component " + name + " not found");
+    }
+
+    mock_component *mock_comp = find(component);
+    if (nullptr == mock_comp) {
+        mock_comp = new mock_component(component, m_mock);
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_select.cc:179]: > Why can't each mock_component contain its own mock instance?

That would be a good idea, if we are 100% sure that component mocks are fully disjoint. But we cannot guarantee that, because different components may point to the same `uct_tl_t` instances, and therefore we may end mocking the same pointer multiple times.
What's wrong with mocking the same pointer twice?
It's a problem to revert mock changes to the original state, for example:
```C++
orig_fun_ptr = &orig_fun;
// remember CURRENT value (&orig_fun), restores on destroy: orig_fun_ptr = &orig_fun;
mock_func(&orig_fun_ptr, mock_fun);
// second mock, same ptr
// remember CURRENT value (&mock_fun), restores on destroy: orig_fun_ptr = &mock_fun;
mock_func(&orig_fun_ptr, mock_fun);
// which one is destroyed first?
```
Now you have a situation when behaviour depends on destructor order. We don't want that. We must use single mock instance for all UCT components

> Or static instance?

Yes, static would work.
However to me it seems that `m_mock` field in fixture is superior alternative, because it's automatically cleaned up upon test case completion. And also I try to avoid global variables when possible, IMO it's better to have explicit dependency injection here.

**Comment by brminich** [comment]
failure in  `rcx/test_ucp_proto_mock.mock_iface_attr/0` looks relevant

**Comment by iyastreb** [comment]
> failure in `rcx/test_ucp_proto_mock.mock_iface_attr/0` looks relevant

Yes, this PR is open for a long time, and new changes in master break test assumptions.
Fixed

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+                    dump_select_elem(worker, select_elem);
+                });
+        }
+    }
+
+    static bool
+    key_match(ucp_proto_select_key_t req, ucp_proto_select_key_t actual)
+    {
+#define CMP_FIELD(FIELD) \
+        if ((UINT8_MAX != req.param.FIELD) && \
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:364]: why comparing to UINT8_MAX?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+            iface_attr.cap.am.max_short = 208;
+            iface_attr.bandwidth.shared = 10000000000;
+            iface_attr.latency.c        = 0.000006;
+            iface_attr.latency.m        = 0.000000001;
+        });
+
+    ucp_test::init();
+    if (!has_resource(sender(), "rc_mlx5", "mlx5_0:1")) {
+        UCS_TEST_SKIP_R("no mlx5_0:1 device");
+    }
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:391]: can we unite it with mock setup?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+        CMP_FIELD(op_attr);
+        CMP_FIELD(dt_class);
+        CMP_FIELD(mem_type);
+        return true;
+    }
+};
+
+UCS_TEST_P(test_ucp_proto_mock, mock_iface_attr, "NET_DEVICES=mlx5_0:1,lo")
+{
+    setup_mock("ib").set_mock_iface_attr("rc_mlx5/mlx5_0:1",
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:379]: can we implement mock so that passing component name is not required?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+public:
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant(variants, UCP_FEATURE_TAG | UCP_FEATURE_AM);
+    }
+
+    virtual void init() override
+    {
+        modify_config("PROTO_ENABLE", "y");
+        modify_config("MAX_RNDV_LANES", "1");
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:223]: why are these needed?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+    {
+        ucp_proto_select_key_t key;
+        key.u64 = UINT64_MAX;
+        return key;
+    }
+
+    void connect()
+    {
+        sender().connect(&receiver(), get_ep_params());
+        for (int i = 0; i < 10; ++i) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:263]: why needed and why 10 is enough?

Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:263]: Right, it's ugly, better wait until certain condition happens, will reimplement this

File: src/ucs/sys/topo/base/topo.h [hunk 21]
```diff
@@ -72,6 +72,10 @@ typedef ucs_status_t
 extern ucs_list_link_t ucs_sys_topo_providers_list;
 
 
+/* Selected topo provider */
+extern struct ucs_sys_topo_provider *ucs_sys_topo_provider;
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply src/ucs/sys/topo/base/topo.h:76]: Ok, I think it's better to have a reset API rather then extending `ucs_sys_topo_get_provider` with reset flag, because the latter would require one extra check in the production flow

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 17 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:391]: No, unfortunately we can't..
Because we must setup mock before ucp_init, in order to intercept all the iface opening and perf estimation functions. So we cannot check sender resources at that point.
But I can implement this skip slightly differently, so that we just call a single function here (`check_mocks`) that makes sure ALL the configured mocks got initialized and skip test otherwise, without explicit call to `has_resource`

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 18 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc]: done

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 22]
```diff
@@ -8,130 +8,80 @@
...
-        uct_tl_t *tl;
-        ucs_list_for_each(tl, &m_component->tl_list, list) {
-            if (0 == strcmp(tl->name, params->mode.device.tl_name)) {
-                return tl;
+        uct_component_h component;
+        ucs_list_for_each(component, &uct_components_list, list) {
+            uct_tl_t *tl;
+            ucs_list_for_each(tl, &component->tl_list, list) {
+                if (tl_name == tl->name) {
+                    EXPECT_EQ(0, m_tls.count(tl_name));
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:60]: why needed?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 23]
```diff
@@ -178,39 +128,27 @@ class mock_component {
         return status;
     }
 
-    uct_component_h                                     m_component;
-    ucs::mock                                          &m_mock;
+    /* We have to use singleton to mock C functions */
+    static mock_iface *singleton;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:132]: maybe m_signleton or m_instance?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 24]
```diff
@@ -219,20 +157,22 @@ class test_ucp_proto_mock : public ucp_test, public mock_component_test {
 
     virtual void init() override
     {
+        /* Reset topo provider to force reload from config */
+        ucs_sys_topo_reset_provider();
+        modify_config("TOPO_PRIO", "default");
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:162]: can u document each env var why is it needed?

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 16 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_proto_mock.cc:289]: ok, can you pls document this in any_key and key_match functions?

Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:332]: sure

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 24 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:204]: sure, done

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 23 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc]: maybe shorter `m_self`?

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 22 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc]: Currently mock works with a single transport, so this is just a check that we don't try to mock multiple ones. In the new impl it's slightly different

**Comment by iyastreb** [comment]
> @iyastreb can you pls revert the refactoring and only fix review comments and refactor in next pr?

Ok, done

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 25]
```diff
@@ -159,8 +124,18 @@ class test_ucp_proto_mock : public ucp_test, public mock_iface {
...
         ucs_sys_topo_reset_provider();
+
+        /*
+         * By default TOPO_PRIO="sysfs,default"
+         * We keep only default config to always have the same topo distances
+         * when test is being executed on different machines.
+         */
         modify_config("TOPO_PRIO", "default");
+
+        /* This test is for dynamic protocol selection available only in v2 */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto_mock.cc:135]: maybe we skill the test altogether if protov2 is disabled? since the default is that it's enabled

Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc]: good idea, done

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 26]
```diff
@@ -0,0 +1,343 @@
...
+        unsigned max_dev_path   = iface_attr->dev_num_paths;
+        (it->second)(*iface_attr);
+
+        /*
+         * It's hard to mock more paths than are actually supported by the
+         * hardware, so in this case we want to skip the test. Returning an
+         * error or skipping the test directly from this point is problematic,
+         * therefore we remember the error and check it in skip_on_error() later
+         */
+        if (iface_attr->dev_num_paths > max_dev_path) {
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_proto_mock.cc:107]: why can't we mock more paths? What is the issue?

File: test/gtest/ucp/test_ucp_proto_mock.cc [hunk 27]
```diff
@@ -0,0 +1,408 @@
...
+        key.u64 = UINT64_MAX;
+        return key;
+    }
+
+    void connect()
+    {
+        sender().connect(&receiver(), get_ep_params());
+        for (int i = 0; i < 10; ++i) {
+            progress();
+        }
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc]: I think this is stale comment, this is fixed already

File: src/ucs/sys/topo/base/topo.c [hunk 28]
```diff
@@ -54,7 +54,7 @@ typedef struct {
 /*
  * Structure needed to define a topology module implementation
  */
-typedef struct {
+typedef struct ucs_sys_topo_provider {
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply src/ucs/sys/topo/base/topo.c]: outdated

File: test/gtest/ucp/test_ucp_proto_mock.cc (hunk 26 above)
Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:107]: There are 2 difficulties:

1. I noticed that dev_num_paths are implicitly "aligned" between IB transports (e.g. between `rc_mlx5` and `ud_mlx5`), in function `ucp_address_gather_devices`. There are 2 possible workarounds for this problem: either we mock both transports (which comes with just another problem: mocking of different TLs is hard to do); or we mock MD query to return some mock MD name for rc_mlx5.
I tried both options, I like more the second one, and by using it we can create as many dev paths as we want

2. wireup tries to connect all the lanes for all the paths, and it fails to do it with mocked paths (Destination is unreachable). It works well when real device supports that amount of paths. But when it exceeds it, then new paths are successfully created (e.g. mock_1:0/path2), but then wireup fails -> connection is terminating -> it hangs forever. Maybe it deserves a bit more investigation, I think we can find solution in the future. For now I just explicitly disable this

Comment (yosefe) [reply test/gtest/ucp/test_ucp_proto_mock.cc:107]: can we force num paths to 1?

Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:107]: We can force it in this PR if you want. But this is a serious limitation, which makes mocking useless for the task I'm working on. Having multiple paths is crucial to reproduce existing bugs in protocol selection:
- issue with lane sorting by BW, which happens only when device with multiple paths is present
- issue with incorrect performance estimation when multiple paths get assigned the entire device performance

Comment (brminich) [reply test/gtest/ucp/test_ucp_proto_mock.cc:107]: maybe you can set `IB_NUM_PATHS` to the value you need in the particular test?

Comment (iyastreb) [reply test/gtest/ucp/test_ucp_proto_mock.cc:107]: Thank you so much for this proposal, it works like a charm!
This is exactly what I needed to mock multiple paths. So I remove all these checks

**Review by brminich — APPROVED** [review]
